Add Serde Integration

l1npengtul opened the pull request:
Adds serde integration, allowing users to use serde with oxilangtag  by using feature `serde`. Bumped to 0.2 in case anyone depended on oxilangtag not supporting serde.

Tpt reviewed (commented):
Thank you for your PR!

I have one "design" concern: do we want to serialize the language tag using the current oxilangtag internal representation or as a string? Using the oxilangtag internal representation exposes oxilangtag internals. This might make interoperability harder by making changes to oxigraph internal representation "breaking" changes and forces other readers of the serialized data to know how to parse this internal representation.

If we want to (de)serialize from/to a string, I believe [Serd derive `try_from`](https://serde.rs/container-attrs.html#try_from) and [Serd derive `into`](https://serde.rs/container-attrs.html#into) are more or less what we want (we would just need to implement `TryFrom<T> for LanguageTag<T>` that is a useful implementation to have anyway).

l1npengtul commented:
Hmm, Ill take these into account. Thank you for the feedback!

l1npengtul commented:
Implemented against the concerns. Look it over once more please
- Added Into String, etc 
- Told users to use serde attrs.
- Added non-falliable version that does this by auto-resolving to en-US
- updated docs, etc

Tpt commented on src/lib.rs:
```diff
@@ -463,6 +467,113 @@ impl<'a> From<LanguageTag<String>> for LanguageTag<Cow<'a, str>> {
+#[cfg(feature = "serde")]
+impl Default for SerdeLanguageTag {
+    fn default() -> Self {
+        match LanguageTag::parse("en-US") {
```
The "und" (undefined) language tag might be better.

l1npengtul replied:
Changed, Thanks!

Tpt reviewed (commented):
Thank you!

After a bit of trial and errors I believe I found a fairly convenient way to implement `Serialize` and `Deserialize` directly on the `LanguageTag` struct with proper validation. I implements directly `Serialize` and `Deserialize`:

```rust
impl<T: Serialize> Serialize for LanguageTag<T> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
       self.tag.serialize(serializer)
    }
}


impl<'de, T: Deref<Target = str> + Deserialize<'de>> Deserialize<'de> for LanguageTag<T> {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<LanguageTag<T>, D::Error> {
        use serde::de::Error;

        LanguageTag::parse(T::deserialize(deserializer)?).map_err(D::Error::custom)
    }
}
```

What do you think about it?

l1npengtul commented:
This looks great! I also didnt know abput the `undefined` tag, thanks for letting me know

Tpt commented on Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "oxilangtag"
-version = "0.1.2"
+version = "0.2.0"
```
It seems to me we are backward compatible so no need to upgrade to 0.2.

l1npengtul replied:
Changed to 0.1.3

Tpt commented on Cargo.toml:
```diff
@@ -13,9 +13,22 @@ Simple and fast implementation of language tag normalization and validation
+
+# enable serde in docs.rs
+[package.metadata.docs.rs]
+features = ["serde"]
```
Maybe just `all-features = true` to be safe if we add other features in the future?

Tpt commented on src/lib.rs:
```diff
@@ -463,14 +471,185 @@ impl<'a> From<LanguageTag<String>> for LanguageTag<Cow<'a, str>> {
+///
+/// See [`LanguageTag`] for details.
+#[derive(Clone, Debug, Hash, Ord, PartialOrd, Eq, PartialEq)]
+pub struct SerdeLanguageTag {
```
I believe this "infallible" language tag parsing is kind of a hack. I am not very happy to include in in `oxilangtag` as a good thing to do. I would be more in favor of leaving it out, people that have a use case for that are able to implement it again anyway. Additionaly they are likely to customize the default tag on parse error, operation that is not possible with the current API.

l1npengtul replied:
Yep, feels like a hack to me as well, I will remove it. Users can implement their own if they wish.

Tpt commented on src/lib.rs:
```diff
@@ -463,14 +471,185 @@ impl<'a> From<LanguageTag<String>> for LanguageTag<Cow<'a, str>> {
-impl fmt::Display for LanguageTagParseError {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+impl Display for LanguageTagParseError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
```
I find it more lenient to have the `fmt::` namespace everywhere to avoid mixing stuff with and without the namespace. But it's bike-shedding so feel free to ignore my comment.

l1npengtul replied:
I got a bunch of errors when I wad compiling, rustc wad complaining that it was an "unnecessary qualifier"

Tpt replied:
It's because `use std::fmt::{Display, Formatter};` at the beginning of the file. If you remove this line, it should not raise any warning.

Tpt commented on tests/lib.rs:
```diff
@@ -696,3 +698,36 @@ fn test_str() {
+fn test_serdelangtag_serialize_deserialize_valid() {
+    let tag: SerdeLanguageTag = LanguageTag::parse("en-US".to_string()).unwrap().into();
+    let serialize_json = serde_json::to_string(&tag).unwrap();
+    println!("{}", serialize_json);
```
No print in tests, please.

l1npengtul replied:
Removed.

Tpt commented on src/lib.rs:
```diff
@@ -463,14 +471,185 @@ impl<'a> From<LanguageTag<String>> for LanguageTag<Cow<'a, str>> {
+    }
+}
+
+impl<T> Into<String> for LanguageTag<T> where T: Deref<Target = str> {
```
I am not sure we need these impls anyway. We should maybe move them out (to an other PR?).

l1npengtul replied:
Since we implemented serialize-deserialize, yes we do not need them anymore. Will remove.

Tpt commented on src/lib.rs:
```diff
@@ -463,14 +471,185 @@ impl<'a> From<LanguageTag<String>> for LanguageTag<Cow<'a, str>> {
+    }
+}
+
+impl TryFrom<String> for LanguageTag<String> {
```
I am not sure we need these impls anyway. We should maybe move them out (to an other PR?).

l1npengtul replied:
ditto from above.

Tpt reviewed (changes requested):
Thank you! It's moving forward. I just have one concern around the "infallible" serde deserialization API.

If you have time, have a look at the CI output, there seems to be some errors in the `Cargo.toml`.

Tpt commented on Cargo.toml:
```diff
@@ -13,9 +13,22 @@ Simple and fast implementation of language tag normalization and validation
 
+[features]
+default = []
+serialize = ["serde/derive"]
```
I believe we don't need Serde `derive` feature anymore because we only "manually" implement `Serialize` and `Deserialize`.
We might only keep serde as an optional dependency and enable `Serialize` and `Deserialize` if it is available.
What do you think about it?

l1npengtul replied:
Oops, forgot about that :p. Thanks, will fix.

Tpt reviewed (commented):
Thank you! It looks great!

I just have a simplication suggestion now that we do not rely on Serde "derive" system anymore.

Tpt commented on Cargo.toml:
```diff
@@ -13,9 +13,22 @@ Simple and fast implementation of language tag normalization and validation
 
+[features]
+default = []
+serialize = ["serde"]
```
I would maybe omit completely the "serialize" feature in favor of just depending on the "sled" dependency being present or not (with just `#[cfg(feature = "serde")]` in the code). This would make usages simpler. Because of trait impls constraints, adding `Serialize` and `Deserialize` implementations should not break any user of this crate because the trait can only be implemented here in oxilangtag or in Serde (where it it isn't...).

l1npengtul replied:
I do not believe I understand what you mean with this. What do you mean by depending on `sled`? Isn't that the KV store database? 

This should not break compatibility, since users can not implement traits onto foreign types.

Tpt replied:
Sorry, I mixed up "sled" and "serde". Sorry for the confusion.

Tpt reviewed (commented):
Thank you! I have one final comment. Feel free to disagree if you think I'm wrong, I am not fully at ease with keeping backward compatiblity in Rust crates.

Tpt commented:
@l1npengtul Is it fine for you if I merge this PR and do some follow-ups myself?

l1npengtul commented:
Sure!

Tpt commented:
Thank you!
